File: lld/problems/chess/go/chess_test.go
package chess

import "testing"

func sq(t *testing.T, s string) Position {
	t.Helper()
	pos, err := ParseSquare(s)
	if err != nil {
		t.Fatalf("bad square %q: %v", s, err)
	}
	return pos
}

func TestPawnOpeningMoveHappyPath(t *testing.T) {
	g := NewGame()
	from, to := sq(t, "e2"), sq(t, "e4")

	if err := g.Move(from, to); err != nil {
		t.Fatalf("expected legal opening move, got err: %v", err)
	}
	if g.Board.At(to) == nil || g.Board.At(to).Type() != Pawn {
		t.Fatalf("expected pawn at e4")
	}
	if g.Board.At(from) != nil {
		t.Fatalf("expected e2 to be empty after move")
	}
	if g.Turn != Black {
		t.Fatalf("expected turn to switch to Black, got %v", g.Turn)
	}
}

func TestIllegalMoveRejected(t *testing.T) {
	g := NewGame()

	// Pawn cannot jump 3 squares.
	if err := g.Move(sq(t, "e2"), sq(t, "e5")); err != ErrIllegalMove {
		t.Fatalf("expected ErrIllegalMove for 3-square pawn move, got %v", err)
	}

	// Rook is blocked by its own pawn at the start of the game.
	if err := g.Move(sq(t, "a1"), sq(t, "a3")); err != ErrIllegalMove {
		t.Fatalf("expected ErrIllegalMove for blocked rook move, got %v", err)
	}
}

func TestCheckDetection(t *testing.T) {
	g := &Game{Board: &Board{}, Turn: White}
	g.Board.set(sq(t, "e1"), NewPiece(King, White))
	g.Board.set(sq(t, "e8"), NewPiece(Rook, Black))
	g.Board.set(sq(t, "a1"), NewPiece(King, Black))

	if !g.IsInCheck(White) {
		t.Fatalf("expected White king on e1 to be in check from rook on e8")
	}
	if g.IsCheckmate(White) {
		t.Fatalf("expected check but not checkmate: king can step aside")
	}
}

// TestFoolsMateCheckmate replays the fastest possible checkmate:
// 1. f3 e5 2. g4 Qh4#
func TestFoolsMateCheckmate(t *testing.T) {
	g := NewGame()
	moves := [][2]string{
		{"f2", "f3"}, // White
		{"e7", "e5"}, // Black
		{"g2", "g4"}, // White
		{"d8", "h4"}, // Black delivers checkmate
	}
	for _, m := range moves {
		if err := g.Move(sq(t, m[0]), sq(t, m[1])); err != nil {
			t.Fatalf("move %s-%s failed: %v", m[0], m[1], err)
		}
	}

	if !g.IsInCheck(White) {
		t.Fatalf("expected White king to be in check after Qh4")
	}
	if !g.IsCheckmate(White) {
		t.Fatalf("expected fool's mate to be checkmate")
	}
}

File: lld/problems/chess/java/src/ChessTest.java
/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out ChessTest` directly.
 */
public class ChessTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testPawnOpeningMoveHappyPath();
        testIllegalMoveRejected();
        testCheckDetection();
        testFoolsMateCheckmate();
        System.out.println("All ChessTest cases passed.");
    }

    private static void testPawnOpeningMoveHappyPath() {
        Game g = new Game();
        Position from = Position.parse("e2");
        Position to = Position.parse("e4");

        g.move(from, to);
        assertTrue(g.getBoard().at(to) != null && g.getBoard().at(to).getType() == PieceType.PAWN,
                "expected pawn at e4");
        assertTrue(g.getBoard().at(from) == null, "expected e2 to be empty after move");
        assertEquals(Color.BLACK, g.getTurn(), "expected turn to switch to Black");
    }

    private static void testIllegalMoveRejected() {
        Game g = new Game();

        try {
            g.move(Position.parse("e2"), Position.parse("e5"));
            throw new AssertionError("expected IllegalMoveException for 3-square pawn move");
        } catch (Game.IllegalMoveException expected) {
            // expected
        }

        try {
            g.move(Position.parse("a1"), Position.parse("a3"));
            throw new AssertionError("expected IllegalMoveException for blocked rook move");
        } catch (Game.IllegalMoveException expected) {
            // expected
        }
    }

    private static void testCheckDetection() {
        Board board = new Board();
        board.set(Position.parse("e1"), PieceFactory.create(PieceType.KING, Color.WHITE));
        board.set(Position.parse("e8"), PieceFactory.create(PieceType.ROOK, Color.BLACK));
        board.set(Position.parse("a1"), PieceFactory.create(PieceType.KING, Color.BLACK));
        Game g = new Game(board, Color.WHITE);

        assertTrue(g.isInCheck(Color.WHITE), "expected White king on e1 to be in check from rook on e8");
        assertTrue(!g.isCheckmate(Color.WHITE), "expected check but not checkmate: king can step aside");
    }

    // Fastest possible checkmate: 1. f3 e5 2. g4 Qh4#
    private static void testFoolsMateCheckmate() {
        Game g = new Game();
        String[][] moves = {
                {"f2", "f3"},
                {"e7", "e5"},
                {"g2", "g4"},
                {"d8", "h4"},
        };
        for (String[] m : moves) {
            g.move(Position.parse(m[0]), Position.parse(m[1]));
        }

        assertTrue(g.isInCheck(Color.WHITE), "expected White king to be in check after Qh4");
        assertTrue(g.isCheckmate(Color.WHITE), "expected fool's mate to be checkmate");
    }

    private static void assertTrue(boolean condition, String label) {
        if (!condition) {
            throw new AssertionError(label);
        }
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
